### PR TITLE
IPAdapterTesterMixin

### LINKED ISSUE #6829: [test] add a `IPAdapterTexterMixin`
I think we should add a `IPAdapterTexterMIxin` and run fast tests on every pipeline that supports IP-Adapter, so that we will be able to catch breaking changes like this https://github.com/huggingface/diffusers/pull/6828

some thoughts:
* we can add a `IPAdapterTexterMIxin` class to `test_pipelines_common.py` https://github.com/huggingface/diffusers/blob/9cc59ba0891c3ccbebfcbb6784f4dde2d5bd8556/tests/pipelines/test_pipelines_common.py#L62C7-L62C32
* we can then add `IPAdapterTexterMIxin` as a base class to the fast tests of all the pipelines support ip-adapter, for example,add here for `StableDiffusionXLPipelineFastTests` (https://github.com/huggingface/diffusers/blob/9cc59ba0891c3ccbebfcbb6784f4dde2d5bd8556/tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py#L59)
* we can either make a dummy repo for ip-adapter ( i.e.`hf-internal-testing/tiny_random_ip_adapter`) or use this method https://github.com/huggingface/diffusers/blob/9cc59ba0891c3ccbebfcbb6784f4dde2d5bd8556/tests/models/unets/test_models_unet_2d_condition.py#L59C34-L59C39 to create `dummy_ip_adapter_state_dict` based on `self.get_dummy_components()['unet']`
*  add a simple test: 
   * the output of the pipeline without ip-adapter should be the same as pipeline with ip-adapter but with ip_adapter_scale = 0
   * the output of pipeline without ip-adapter should NOT be same as pipeline with ip-adapter with ip_adapter_scale = 1
* the tests should only run if the pipeline's  `__call__` method has the `ip_adapter_image` argument

cc @DN6 @sayakpaul let me know what you guys think!

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Adds IPAdapterTesterMixin to ensure pipelines supporting it are not broken with newer changes.

Fixes #6829.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@yiyixuxu @sayakpaul

### CODE REVIEW

File: tests/pipelines/test_pipelines_common.py [hunk 1]
```diff
@@ -59,6 +61,107 @@ def check_same_shape(tensor_list):
     return all(shape == shapes[0] for shape in shapes[1:])
 
 
+class IPAdapterTesterMixin:
```
Comment (a-r-r-o-w) [inline tests/pipelines/test_pipelines_common.py:64]: Please let me know your thoughts on if this has been done as you expected. Pipelines that support IPAdapter's should derive from this and create dummy image_encoder and feature_extractor instead of `None`.

**Comment by yiyixuxu** [comment]
thanks so much for working on this!!! 😇😇😇
we are discussing about allowing pass image embedding as `ip_adapter_image` here https://github.com/huggingface/diffusers/discussions/6830 I think it will make the testing a lot easier too, no? that way we don't need to make an `image_encoder` for all the pipelines. Let me know what you think!

**Comment by a-r-r-o-w** [comment]
Yes that makes sense since `ip_adapter_image` is annotated as `PipelineImageInput`, which can be a PIL.Image, np.ndarray or torch.Tensor, and should've maybe been the default behaviour to accept anything with appropriate validation from the start. Once a PR supporting embeddings is merged, I'll update the relevant testing code here. Also looks like all tests passed. Do you want me to add the IPAdapterTesterMixin to all the pipeline tests where relevant (like SD/SDXL I2I, Inpaint)? I think it might be too redundant and makes sense to have only for T2I since most, if not all, other implementations use the `# Copied from` logic.

Btw, am I right in assuming that IP-Adapter-Plus tests should also be added here?

File: tests/pipelines/pipeline_params.py [hunk 2]
```diff
@@ -127,3 +127,5 @@
 TEXT_TO_IMAGE_CALLBACK_CFG_PARAMS = frozenset(["prompt_embeds"])
 
 VIDEO_TO_VIDEO_BATCH_PARAMS = frozenset(["prompt", "negative_prompt", "video"])
+
+IP_ADAPTER_PARAMS = frozenset(["ip_adapter_image"])
```
Comment (yiyixuxu) [inline tests/pipelines/pipeline_params.py:131]: I think we don't need this for now - so far all pipelines use `ip_adapter_image`

File: tests/pipelines/stable_diffusion/test_stable_diffusion.py [hunk 3]
```diff
@@ -154,6 +172,15 @@ def get_dummy_components(self, time_cond_proj_dim=None):
         )
         text_encoder = CLIPTextModel(text_encoder_config)
         tokenizer = CLIPTokenizer.from_pretrained("hf-internal-testing/tiny-random-clip")
+        feature_extractor = CLIPImageProcessor()
```
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion/test_stable_diffusion.py:175]: once this PR https://github.com/huggingface/diffusers/pull/6868 is merged, IMO we can maybe skip this step ( i.e. creating `feature_extractor` and `image_encoder`) and test with `ip_adapter_image_embedding` directly. 
I think it is ok to test with image embeddings directly because the process to convert `ip_adapter_image` to `ip_adapter_image_embedding` is the same across all pipelines, using the same `prepare_ip_adapter_image_embeds` method; and It will be a lot easier to add this IP adapter test to a pipeline without this step.

but ok to keep it here for now and we can discussion more:)

File: tests/pipelines/test_pipelines_common.py [hunk 4]
```diff
@@ -59,6 +61,107 @@ def check_same_shape(tensor_list):
...
+    def test_image_encoder_exists(self):
+        components = self.get_dummy_components()
+        self.assertIsNotNone(components.get("image_encoder", None))
+
+    def test_pipeline_signature(self):
+        parameters = inspect.signature(self.pipeline_class.__call__).parameters
+
+        assert issubclass(self.pipeline_class, IPAdapterMixin)
+        for param in self.ip_adapter_params:
+            self.assertIn(param, parameters, "`ip_adapter_image` argument must be supported by the `__call__` method")
```
Comment (yiyixuxu) [inline tests/pipelines/test_pipelines_common.py:86]:
```suggestion
        self.assertIn("ip_adapter_image" , parameters, "`ip_adapter_image` argument must be supported by the `__call__` method")
```

File: tests/pipelines/test_pipelines_common.py [hunk 5]
```diff
@@ -59,6 +61,107 @@ def check_same_shape(tensor_list):
...
+            max_diff_with_multi_adapter_scale,
+            0.1,
+            "Output with multi-ip-adapter scale must be different from normal inference",
+        )
+
+    def test_ip_adapter_plus(self):
+        # TODO
+        pass
+
+
```
Comment (yiyixuxu) [inline tests/pipelines/test_pipelines_common.py:164]:
```suggestion
```
I think we don't need to test all types here:)

**Review by yiyixuxu — COMMENTED** [review]
hey thanks! looks really good!
let's wait a little bit for https://github.com/huggingface/diffusers/pull/6868 to merge and we can finish it up!

**Comment by yiyixuxu** [comment]
> Do you want me to add the IPAdapterTesterMixin to all the pipeline tests where relevant (like SD/SDXL I2I, Inpaint)? I think it might be too redundant and makes sense to have only for T2I since most, if not all, other implementations use the # Copied from logic.

I think we should add to all pipeline that support ip-adapter; but we should only use IP adapter image embedding directly, so the code addition to the fast tests is actually minimum; I think you just have to append the mixin to the test that's all :)

File: tests/pipelines/stable_diffusion/test_stable_diffusion.py (hunk 3 above)
Comment (a-r-r-o-w) [reply tests/pipelines/stable_diffusion/test_stable_diffusion.py:175]: yes that works and will be a simpler addition with just needing to inherit this mixin to test a pipeline

File: tests/pipelines/test_pipelines_common.py [hunk 6]
```diff
@@ -59,6 +61,93 @@ def check_same_shape(tensor_list):
...
+    """
+    This mixin is designed to be used with PipelineTesterMixin and unittest.TestCase classes.
+    It provides a set of common tests for pipelines that support IP Adapters.
+    """
+
+    def test_pipeline_signature(self):
+        parameters = inspect.signature(self.pipeline_class.__call__).parameters
+
+        assert issubclass(self.pipeline_class, IPAdapterMixin)
+        self.assertIn(
```
Comment (yiyixuxu) [inline tests/pipelines/test_pipelines_common.py:74]: let's make sure `ip_adapter_image` is a parameter too

**Review by yiyixuxu — COMMENTED** [review]
love this! thank you!

**Comment by yiyixuxu** [comment]
look, it already caught a bunch of test failures!!! 😅
I think most of them have the same type of error so probably easy to fix. 

Not all of the pipelines need to have ip-adapter support IMO. So if the pipeline isn't very much used, e.g. stable diffusion safe, and the test failure is because the ip-adapter isn't correctly implemented. Let me know! we can probably remove it

**Comment by a-r-r-o-w** [comment]
Thanks for the Good PR label :laughing: Just adding the IPAdapterTesterMixin to the class defintion was quite convenient!

I agree with removing the tests from the not-so-used pipelines that are not regularly discussed about in issues. I will remove them from what I think are lesser used, but since you have access to make changes to my branch, please feel free to change it to your liking.

File: tests/pipelines/animatediff/test_animatediff_video2video.py [hunk 7]
```diff
@@ -28,7 +28,7 @@ def to_np(tensor):
     return tensor
 
 
-class AnimateDiffVideoToVideoPipelineFastTests(PipelineTesterMixin, unittest.TestCase):
+class AnimateDiffVideoToVideoPipelineFastTests(IPAdapterTesterMixin, PipelineTesterMixin, unittest.TestCase):
```
Comment (a-r-r-o-w) [inline tests/pipelines/animatediff/test_animatediff_video2video.py:31]: This pipeline will fail tests because it uses the old logic for IPAdapters with `encode_image()` instead of `prepare_ip_adapter_image_embeds`. Can I fix the IP Adapter support here? Or should it be done in a different PR?

File: tests/pipelines/stable_diffusion_sag/test_stable_diffusion_sag.py [hunk 8]
```diff
@@ -32,13 +32,15 @@
...
 from ..pipeline_params import TEXT_TO_IMAGE_BATCH_PARAMS, TEXT_TO_IMAGE_IMAGE_PARAMS, TEXT_TO_IMAGE_PARAMS
-from ..test_pipelines_common import PipelineLatentTesterMixin, PipelineTesterMixin
+from ..test_pipelines_common import IPAdapterTesterMixin, PipelineLatentTesterMixin, PipelineTesterMixin
 
 
 enable_full_determinism()
 
 
-class StableDiffusionSAGPipelineFastTests(PipelineLatentTesterMixin, PipelineTesterMixin, unittest.TestCase):
+class StableDiffusionSAGPipelineFastTests(
```
Comment (a-r-r-o-w) [inline tests/pipelines/stable_diffusion_sag/test_stable_diffusion_sag.py:41]: This pipeline will also fail tests because of old IP Adapter logic with `encode_image`. I think SAG is used a bit in community (atleast I make use of it sometimes), so it makes sense to support IP Adapters with latest implementation and tests here. WDYT? @yiyixuxu

File: tests/pipelines/latent_consistency_models/test_latent_consistency_models.py [hunk 9]
```diff
@@ -20,13 +20,15 @@
...
 from ..pipeline_params import TEXT_TO_IMAGE_BATCH_PARAMS, TEXT_TO_IMAGE_IMAGE_PARAMS, TEXT_TO_IMAGE_PARAMS
-from ..test_pipelines_common import PipelineLatentTesterMixin, PipelineTesterMixin
+from ..test_pipelines_common import IPAdapterTesterMixin, PipelineLatentTesterMixin, PipelineTesterMixin
 
 
 enable_full_determinism()
 
 
-class LatentConsistencyModelPipelineFastTests(PipelineLatentTesterMixin, PipelineTesterMixin, unittest.TestCase):
+class LatentConsistencyModelPipelineFastTests(
```
Comment (a-r-r-o-w) [inline tests/pipelines/latent_consistency_models/test_latent_consistency_models.py:29]: Tests will fail due to old implementation of IP Adapters both here and in lcm img2img

**Comment by a-r-r-o-w** [comment]
Pipelines that have the error that looks something like:
```py
FAILED tests/pipelines/controlnet/test_controlnet_img2img.py::ControlNetImg2ImgPipelineFastTests::test_ip_adapter - TypeError: argument of type 'NoneType' is not iterable
FAILED tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py::StableDiffusionImg2ImgPipelineFastTests::test_ip_adapter - TypeError: argument of type 'NoneType' is not iterable
```
is because the IP Adapter implementation is incorrect:
```py
added_cond_kwargs = {"image_embeds": image_embeds} if ip_adapter_image is not None else None
```
The check should include both ip_adapter_image and ip_adapter_image_embeds as done in some pipelines of #6868 but looks like a few spots were missed. Looks like this mixin will help make things quite consistent!

File: tests/pipelines/animatediff/test_animatediff_video2video.py (hunk 7 above)
Comment (yiyixuxu) [reply tests/pipelines/animatediff/test_animatediff_video2video.py:31]: I think it's easier to fix it here:)

File: tests/pipelines/stable_diffusion_sag/test_stable_diffusion_sag.py (hunk 8 above)
Comment (yiyixuxu) [reply tests/pipelines/stable_diffusion_sag/test_stable_diffusion_sag.py:41]: yes I agree :)

File: src/diffusers/pipelines/latent_consistency_models/pipeline_latent_consistency_text2img.py [hunk 10]
```diff
@@ -590,6 +625,10 @@ def cross_attention_kwargs(self):
     def clip_skip(self):
         return self._clip_skip
 
+    @property
+    def do_classifier_free_guidance(self):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/latent_consistency_models/pipeline_latent_consistency_text2img.py:629]: ok :)

**Review by yiyixuxu — APPROVED** [review]
thank you!

**Comment by a-r-r-o-w** [comment]
```
FAILED tests/pipelines/controlnet/test_controlnet.py::ControlNetPipelineFastTests::test_ip_adapter - AssertionError: 0.007904649 not greater than 0.01 : Output with multi-ip-adapter scale must be different from normal inference
FAILED tests/pipelines/controlnet/test_controlnet_inpaint.py::ControlNetSimpleInpaintPipelineFastTests::test_ip_adapter - AssertionError: 0.0 not greater than 0.01 : Output with ip-adapter must be different from normal inference
FAILED tests/pipelines/controlnet/test_controlnet_img2img.py::ControlNetImg2ImgPipelineFastTests::test_ip_adapter - AssertionError: 8.6426735e-07 not greater than 0.01 : Output with ip-adapter must be different from normal inference
FAILED tests/pipelines/stable_diffusion_panorama/test_stable_diffusion_panorama.py::StableDiffusionPanoramaPipelineFastTests::test_ip_adapter - AssertionError: 0.0014307499 not greater than 0.01 : Output with ip-adapter must be different from normal inference
```

@yiyixuxu I'm not sure why these tests fail. I think the results are slightly different based on device because they do pass on my system with a difference greater than 0.01. Is lowering the threshold the ideal solution or what do you have in mind? Maybe a threshold with the differences accumulated could also be used. Regarding ControlNetSimpleInpaintPipelineFastTests, I'm not sure why the outputs are the same.

File: src/diffusers/pipelines/latent_consistency_models/pipeline_latent_consistency_text2img.py (hunk 10 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/latent_consistency_models/pipeline_latent_consistency_text2img.py:629]: just for consistency with the `# Copied from` style - it didn't make sense why we'd want to modify the `prepare_ip_adapter_image_embeds` method for one pipeline instead of reusing the same across all pipelines

File: tests/pipelines/test_pipelines_common.py [hunk 11]
```diff
@@ -59,6 +61,103 @@ def check_same_shape(tensor_list):
...
+        # forward pass with single ip adapter, but scale=0 which should have no effect
+        inputs = self._modify_inputs_for_ip_adapter_test(self.get_dummy_inputs(torch_device))
+        inputs["ip_adapter_image_embeds"] = [self._get_dummy_image_embeds(cross_attention_dim)]
+        pipe.set_ip_adapter_scale(0.0)
+        output_without_adapter_scale = pipe(**inputs)[0]
+
+        # forward pass with single ip adapter, but with scale of adapter weights
+        inputs = self._modify_inputs_for_ip_adapter_test(self.get_dummy_inputs(torch_device))
+        inputs["ip_adapter_image_embeds"] = [self._get_dummy_image_embeds(cross_attention_dim)]
+        pipe.set_ip_adapter_scale(1.0)
```
Comment (yiyixuxu) [inline tests/pipelines/test_pipelines_common.py:117]: Maybe increase this value to 100 lol

**Comment by yiyixuxu** [comment]
@a-r-r-o-w 
for the failing test, maybe trying to increase the scale, see https://github.com/huggingface/diffusers/pull/6862/files#r1491365749

**Comment by a-r-r-o-w** [comment]
Thanks @yiyixuxu. I believe all tests should pass now and that this is ready for a merge.

**Comment by yiyixuxu** [comment]
Hi @a-r-r-o-w 
there is still one test fail for panorama pipeline
can you look into it? I'm ok with removing ip-adapter support from it if it's not easy to get it to work:)

File: tests/pipelines/test_pipelines_common.py [hunk 12]
```diff
@@ -59,6 +61,105 @@ def check_same_shape(tensor_list):
...
+        )
+
+    def _get_dummy_image_embeds(self, cross_attention_dim: int = 32):
+        return torch.randn((2, 1, cross_attention_dim), device=torch_device)
+
+    def _modify_inputs_for_ip_adapter_test(self, inputs: Dict[str, Any]):
+        inputs["output_type"] = "np"
+        inputs["return_dict"] = False
+        if "image" in inputs.keys():
+            inputs["num_inference_steps"] = 4
```
Comment (DN6) [inline tests/pipelines/test_pipelines_common.py:92]: Just a question. Why increase this based on the whether `image` is provided? Because of `strength`?

File: tests/pipelines/test_pipelines_common.py [hunk 13]
```diff
@@ -59,6 +61,105 @@ def check_same_shape(tensor_list):
...
+        return inputs
+
+    def test_ip_adapter(self, expected_max_diff: float = 1e-4):
```
Comment (DN6) [inline tests/pipelines/test_pipelines_common.py:95]: Can we break this up into individually named tests? It is hard to debug when an intermediate assert causes the entire test to fail.

File: tests/pipelines/test_pipelines_common.py (hunk 12 above)
Comment (a-r-r-o-w) [reply tests/pipelines/test_pipelines_common.py:92]: That is correct. The tests fail for img2img tests that have num_inference_steps set to 2 or lower and strength greater than 5. I think if it was:
```py
if "image" in inputs().keys() and "strength" in inputs().keys():
    ...
```
it would make more sense.

**Comment by a-r-r-o-w** [comment]
> Hi @a-r-r-o-w there is still one test fail for panorama pipeline can you look into it? I'm ok with removing ip-adapter support from it if it's not easy to get it to work:)

Yeah, I think I've been having troubles on getting it to work. I'm not sure why the results vary :/

File: tests/pipelines/test_pipelines_common.py (hunk 13 above)
Comment (a-r-r-o-w) [reply tests/pipelines/test_pipelines_common.py]: I've split it into two tests for the time being but note that if we have K classes that derive from this mixin, we now have K additional inferences running as compared to before. The time to execute all tests doesn't increase too much though.

**Comment by a-r-r-o-w** [comment]
@yiyixuxu Apologies for the delay. With the last commit, all tests pass for me locally and I hope this is completed :smiling_face_with_tear:

**Comment by yiyixuxu** [comment]
thank you:)
